embassy-rp/src/dma.rs

```diff
@@ -1,82 +1,234 @@
+
+    irq.enable();
+}
 
-            p.read_addr().write_value(from.as_ptr() as u32);
-            p.write_addr().write_value(to.as_mut_ptr() as u32);
-            p.trans_count().write_value(from.len() as u32);
+pub unsafe fn read<'a, C: Channel, W: Word>(
+    ch: impl Peripheral<P = C> + 'a,
+    from: *const W,
+    to: &mut [W],
+) -> Transfer<'a, C> {
+    let (ptr, len) = crate::dma::slice_ptr_parts_mut(to);
+    copy_inner(ch, from as *const u32, ptr as *mut u32, len, W::size(), false, true)
+}
 
-            compiler_fence(Ordering::SeqCst);
+pub unsafe fn write<'a, C: Channel, W: Word>(
+    ch: impl Peripheral<P = C> + 'a,
+    from: &[W],
+    to: *mut W,
+) -> Transfer<'a, C> {
+    let (from_ptr, len) = crate::dma::slice_ptr_parts(from);
+    copy_inner(ch, from_ptr as *const u32, to as *mut u32, len, W::size(), true, false)
+}
 
-            p.ctrl_trig().write(|w| {
-                w.set_data_size(vals::DataSize::SIZE_WORD);
-                w.set_incr_read(true);
-                w.set_incr_write(true);
-                w.set_chain_to(inner.number());
-                w.set_en(true);
-            });
+pub unsafe fn copy<'a, C: Channel, W: Word>(
+    ch: impl Peripheral<P = C> + 'a,
+    from: &[W],
+    to: &mut [W],
+) -> Transfer<'a, C> {
+    let (from_ptr, from_len) = crate::dma::slice_ptr_parts(from);
+    let (to_ptr, to_len) = crate::dma::slice_ptr_parts_mut(to);
+    assert_eq!(from_len, to_len);
+    copy_inner(
+        ch,
+        from_ptr as *const u32,
+        to_ptr as *mut u32,
+        from_len,
+        W::size(),
+        true,
+        true,
+    )
+}
 
-            while p.ctrl_trig().read().busy() {}
+fn copy_inner<'a, C: Channel>(
+    ch: impl Peripheral<P = C> + 'a,
+    from: *const u32,
+    to: *mut u32,
+    len: usize,
+    data_size: DataSize,
+    incr_read: bool,
+    incr_write: bool,
+) -> Transfer<'a, C> {
+    into_ref!(ch);
+
+    unsafe {
+        let p = ch.regs();
+
+        p.read_addr().write_value(from as u32);
+        p.write_addr().write_value(to as u32);
+        p.trans_count().write_value(len as u32);
+
+        compiler_fence(Ordering::SeqCst);
+
+        p.ctrl_trig().write(|w| {
+            w.set_data_size(data_size);
+            w.set_incr_read(incr_read);
+            w.set_incr_write(incr_write);
+            w.set_chain_to(ch.number());
+            w.set_en(true);
+        });
+
+        compiler_fence(Ordering::SeqCst);
+    }
+    Transfer::new(ch)
+}
 
-            compiler_fence(Ordering::SeqCst);
+pub struct Transfer<'a, C: Channel> {
+    channel: PeripheralRef<'a, C>,
+}
+
+impl<'a, C: Channel> Transfer<'a, C> {
+    pub(crate) fn new(channel: impl Peripheral<P = C> + 'a) -> Self {
+        into_ref!(channel);
+
+        Self { channel }
+    }
+}
+
+impl<'a, C: Channel> Drop for Transfer<'a, C> {
+    fn drop(&mut self) {
+        let p = self.channel.regs();
+        unsafe {
+            p.ctrl_trig().write(|w| w.set_en(false));
```

Comment:
you have to use CHAN_ABORT to stop a channel, setting EN=false isn't enough, see datasheet `2.5.5.3. Channel Abort`

Reply:
Nice catch!